From 80fc6e6a9686ef1e62491b3baf842ad86964500d Mon Sep 17 00:00:00 2001
From: xianhaox <xianhaox.qing@intel.com>
Date: Mon, 2 Apr 2018 17:17:49 +0800
Subject: [PATCH] Add user checking when resume playing back

There is no user checking when restoring audio focus to app. If end user
switched to a new user2 after launched 2 media apps and requested audio
focus in both apps in user1. Then try to request and abandon audio focus
in new user2. The media app in user1 will get audio focus. If the app
does not handle the focus properly then we got the issue.

This solution is to add user checking when trying to resume the play.

Test steps:
    1.Owner:launch Local media player>>play a song
    2.Owner:Launch Radio
    3.Switch to New User >> launch Local media player >> play a song
      >> pause the song >> observe

The background music should not start playing.

Then go to CarSettings adjust the volume and observe.
There should be no crash of LocalMediaPlayer.

Change-Id: I16693dc261db78ad5fc31d232cbfd39e46dcb54f
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-60858
Signed-off-by: xianhaox <xianhaox.qing@intel.com>
---
 AndroidManifest.xml                                    | 2 ++
 src/com/android/car/media/localmediaplayer/Player.java | 8 ++++++++
 2 files changed, 10 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 49e0d35..d080593 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -23,6 +23,8 @@
         android:targetSdkVersion="23" />
 
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.INTERACT_ACROSS_USERS_FULL"/>
+    <uses-permission android:name="android.permission.MANAGE_USERS"/>
 
     <application
         android:allowBackup="true"
diff --git a/src/com/android/car/media/localmediaplayer/Player.java b/src/com/android/car/media/localmediaplayer/Player.java
index f70e05c..0eb65fa 100644
--- a/src/com/android/car/media/localmediaplayer/Player.java
+++ b/src/com/android/car/media/localmediaplayer/Player.java
@@ -15,6 +15,7 @@
  */
 package com.android.car.media.localmediaplayer;
 
+import android.app.ActivityManager;
 import android.app.Notification;
 import android.app.NotificationManager;
 import android.app.PendingIntent;
@@ -333,12 +334,19 @@ public class Player extends MediaSession.Callback {
             mSession.setPlaybackState(mErrorState);
         }
     }
+    boolean userIsInForeground(Context context) {
+        return ActivityManager.getCurrentUser() == context.getUserId();
+    }
 
     private void resumePlayback() {
         if (Log.isLoggable(TAG, Log.DEBUG)) {
             Log.d(TAG, "resumePlayback()");
         }
 
+        if (!userIsInForeground(mContext)) {
+            return;
+        }
+
         updatePlaybackStatePlaying();
 
         if (!mMediaPlayer.isPlaying()) {
-- 
1.9.1

